### PR TITLE
Weights+Biases and ResNet integration

### PR DESCRIPTION
**Pull request by awindler**

### CODE REVIEW

File: configs/example_sweep_config.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  "optimizer": {
+    "value": "keras.optimizers.Adam"
+  },
+  "epochs": {
+    "value": 3
+  },
+  "models": {
+    "values": ["FCN", "ResNet"]
+  }
+}
```
Comment (tbohne) [inline configs/example_sweep_config.py]:
```suggestion
}

```

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,6 @@ data/
 .idea/
 __pycache__/
 *.pyc
-best_model.h5
\ No newline at end of file
+best_model.h5
+configs/api_key.py
+wandb/
```
Comment (tbohne) [inline .gitignore]:
```suggestion
wandb/

```

File: models.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    conv3 = keras.layers.ReLU()(conv3)
+
+    gap = keras.layers.GlobalAveragePooling1D()(conv3)
+
+    output_layer = keras.layers.Dense(num_classes, activation="softmax")(gap)
+
+    return keras.models.Model(inputs=input_layer, outputs=output_layer)
+
+
+def create_resnet_model(input_shape, num_classes):
```
Comment (tbohne) [inline models.py:34]:
```suggestion
def create_resnet_model(input_shape, num_classes):
    """
    Defines the ResNet architecture to be worked with.
    :param input_shape: shape of the input layer
    :param num_classes: number of unique classes to be considered
    :return: ResNet model
    """
```

File: models.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    gap_layer = keras.layers.GlobalAveragePooling1D()(output_block_3)
+
+    output_layer = keras.layers.Dense(num_classes, activation='softmax')(gap_layer)
+
+    model = keras.models.Model(inputs=input_layer, outputs=output_layer)
+
+    return model
+
+
+def create_model(input_shape, num_classes, architecture="FCN"):
```
Comment (tbohne) [inline models.py:115]:
```suggestion
def create_model(input_shape, num_classes, architecture="FCN"):
    """
    Initiates model generation based on the specified architecture (or default).
    :param input_shape: shape of the input layer
    :param num_classes: number of unique classes to be considered
    :return: created model
    """
```

File: models.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    if architecture == "FCN":
+        return create_fcn_model(input_shape, num_classes)
+    elif architecture == "ResNet":
+        return create_resnet_model(input_shape, num_classes)
+    else:
+        raise ValueError("Unknown model architecture: "+architecture)
```
Comment (tbohne) [inline models.py]:
```suggestion
        raise ValueError("Unknown model architecture: " + architecture)

```

File: run_sweep.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    parser = argparse.ArgumentParser(description='Hyperparametertunig with Weights&biases sweep')
+    parser.add_argument('--train_path', type=file_path, required=True)
+    parser.add_argument('--val_path', type=file_path, required=True)
+    parser.add_argument('--test_path', type=file_path, required=True)
+    args = parser.parse_args()
+
+    os.environ["WANDB_API_KEY"] = api_key.wandb_api_key
+    sweep_config = {'method': TUNING_METHOD, 'parameters': example_sweep_config.sweep_config}
+    sweep_id = wandb.sweep(sweep_config, project="Oscillogram Classification CNN")
```
Comment (tbohne) [inline run_sweep.py]:
```suggestion
    sweep_id = wandb.sweep(sweep_config, project="Oscillogram Classification")
```

File: run_sweep.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+from train import file_path
+from configs import api_key, example_sweep_config
+
+TUNING_METHOD = 'random'
+N_RUNS_IN_SWEEP = 3
+
+def main():
+
+    def call_training_procedure(config_dict=None):
```
Comment (tbohne) [inline run_sweep.py:14]: Why did you define this function inside of `main`?

File: train.py [hunk 8]
```diff
@@ -8,11 +8,17 @@
...
+from wandb.keras import WandbCallback
 
 from training_data import TrainingData
+from configs import api_key, example_run_config
+import models
 
-EPOCHS = 50
-BATCH_SIZE = 32
+
+def set_up_wandb(wandb_config):
```
Comment (tbohne) [inline train.py:19]:
```suggestion
def set_up_wandb(wandb_config):
    """
    Setup for 'weights and biases'.
    :param wandb_config: configuration to be used
    """
```

File: train.py [hunk 9]
```diff
@@ -194,6 +175,12 @@ def evaluate_model(test_data_path):
     evaluate_model_on_test_data(x_test.astype('float32'), y_test.astype('float32'))
 
 
+def train_procedure(train_path, val_path, test_path, hyperparameter_config=example_run_config.hyperparameter_config):
```
Comment (tbohne) [inline train.py:182]:
```suggestion
def train_procedure(train_path, val_path, test_path, hyperparameter_config=example_run_config.hyperparameter_config):
    """
    Initiates the training procedure.
    :param train_path: path to training data
    :param val_path: path to validation data
    :param test_path: path to test data
    :param hyperparameter_config: hyper parameter specification
    """
```

**Review by tbohne — COMMENTED** [review]
add review

File: run_sweep.py (hunk 7 above)
Comment (awindler) [reply run_sweep.py:14]: wandb.agent requires a training function which takes only one argument (the hyperparameter config dictionary). Therefore, train.train_procedure cannot be used directly, as it expects multiple arguments. In order to pass args.train_path, args.val_path and args.test_path to train.train_procedure and at the same time have a function that takes only one argument, I defined this inner function. An alternative would be to put this function outside of main and to put the argparse part inside this function.

**Review by tbohne — APPROVED** [review]

**Pull request merged**